File: project/two_pianos/lib/two_pianos_web/channels/lobby_channel.ex
defmodule TwoPianosWeb.LobbyChannel do

  use TwoPianosWeb, :channel

  def join("lobby", _, socket) do
    {:ok, socket}
  end

  def handle_in("create_room", _, socket) do
    # Create protected room (Generate unique room ID and code)
    {room_id, room_code} = RoomManager.create_room(:protected, socket.assigns[:user_id])

    # reply w/ room id and room code -> they will join the room's specific channel, will be
    # able to see the code, and will have a piano to mess around with
    {:reply, {:room_created, %{room_id: room_id, room_code: room_code}}, socket}
  end

  def handle_in("join_existing_room", %{"room_code" => code}, socket) do

    case RoomManager.get_id_by_code(code) do
      nil ->
        {:reply, {:invalid_code, %{}}, socket}
      room_id  ->
        {:reply, {:valid_code, %{room_id: room_id}}}
    end

  end

  def handle_in("match_with_stranger", _, socket) do

    case RandomUserMatcher.match(socket.assigns[:user_id]) do
      :waiting ->
        {:reply, {:waiting, %{}}, socket}
      {:match, matched_user_id} ->
        room_id = RoomManager.create_room(:match, {socket.assigns[:user_id], matched_user_id})
        TwoPianosWeb.UserChannel.broadcast_match(room_id, socket.assigns[:user_id], matched_user_id)
        {:reply, {:match, %{room_id: room_id}}, socket}
    end

  end

end

File: project/two_pianos/assets/js/lobby_channel.js
export class LobbyChannel{

    static setup_and_join(socket, user_id){

        let lobby_channel = socket.channel("lobby", {user_id: user_id})

        // Create New Room button press
        document.getElementById("create-btn").addEventListener("click", e => {

            lobby_channel.push("create_room")
                .receive("room_created", (resp) => {

                    console.log(`Lobby Channel: Room Created\nroom_id: ${resp.room_id}, room_code: ${resp.room_code}`)

                    localStorage.setItem("room_id", resp.room_id)

                    setTimeout(function(){
                        window.location.replace("http://localhost:4000/room")
                    }, 2000)

                })

        })
        
        // Join Room button press
        document.getElementById("join-existing-btn").addEventListener("click", e => {

            let room_code = document.getElementById("room-code-input").value

            lobby_channel.push("join_existing_room", {room_code: room_code})
                .receive("invalid_code", (resp) => {

                    console.log(`Lobby Channel: Join Existing\ninvalid code ${room_code}`)

                })
                .receive("valid_code", (resp) => {

                    console.log(`Lobby Channel: Join Existing\nvalid code ${room_code}, connecting to room ${resp.room_id}`)

                    localStorage.setItem("room_id", resp.room_id)

                    setTimeout(function(){
                        window.location.replace("http://localhost:4000/room")
                    }, 2000)

                })

        })

        // Join Stranger button press
        document.getElementById("match-stranger-btn").addEventListener("click", e => {

            lobby_channel.push("match_with_stranger", {test: "test"})
                .receive("waiting", (resp) => {

                    console.log("Lobby Channel: Match With Stranger\nWaiting")

                })
                .receive("already_waiting", (resp) => {

                    console.log("Lobby Channel: Match With Stranger\nAlready Waiting")

                })
                .receive("match", (resp) => {

                    console.log(`Lobby Channel: Match With Stranger\nMatch, room_id: ${resp.room_id}`)

                    localStorage.setItem("room_id", resp.room_id)

                    setTimeout(function(){
                        window.location.replace("http://localhost:4000/room")
                    }, 2000)

                })
        })

        lobby_channel.join()

        return lobby_channel

    }

}